# บทที่ 8: การนำเข้าไฟล์ข้อมูลด้วย pandas

หลังจากเราได้เรียนเรื่อง `Series` และ `DataFrame` ซึ่งเป็น data structure สำคัญของ pandas สำหรับงานวิเคราะห์ข้อมูล และได้สร้าง DataFrame จาก `dictionary` ด้วยตนเอง แต่ในการทำงานจริง นักวิเคราะห์ข้อมูลมักได้รับข้อมูลจากหลายแหล่ง เช่น 
- ไฟล์ CSV 
- ไฟล์ Excel 
- ไฟล์ text 
- ไฟล์ที่ดาวน์โหลดจากเว็บไซต์

เป้าหมายของบทนี้คือการอ่านไฟล์ให้ถูกต้อง และแปลงข้อมูลจากแหล่งต่าง ๆ ให้กลายเป็น DataFrame

## ผลการเรียนรู้ที่คาดหวัง 
เมื่อเรียนจบบทนี้ ผู้เรียนจะสามารถ 
1. อธิบายได้ว่าการนำเข้าไฟล์ข้อมูลเป็นขั้นตอนเริ่มต้นของ workflow งานวิเคราะห์ข้อมูล 
2. ใช้ `pd.read_csv()` เพื่อนำเข้าไฟล์ CSV เป็น DataFrame ได้ 
3. ใช้ `pd.read_excel()` เพื่อนำเข้าไฟล์ Excel เป็น DataFrame ได้ 
4. ใช้ `pd.read_csv()` เพื่อนำเข้าข้อมูลจาก URL ได้ 
5. อ่านไฟล์ `.txt` ที่มีตัวคั่นข้อมูลเป็น `|` ได้ 
6. ใช้ `encoding` ให้เหมาะสมกับไฟล์ภาษาไทยได้ 
7. ใช้ `sep` เพื่อกำหนดตัวคั่นข้อมูลได้ 
8. ใช้ `header=None` เมื่อไฟล์ไม่มีชื่อ column ได้ 
9. สร้างรายชื่อ column เองด้วย `list` และ `for` loop ได้ 
10. ใช้ `names` เพื่อกำหนดชื่อ column ให้ DataFrame ตอนนำเข้าไฟล์ได้ 
11. ใช้ `usecols` เพื่อเลือกอ่านเฉพาะ column ที่ต้องการได้ 
12. ใช้ `dtype` เพื่อควบคุมชนิดข้อมูลของ column ที่เป็นรหัสได้ 
13. ใช้ `nrows` เพื่อทดลองอ่านไฟล์เพียงบางส่วนได้ 
14. ใช้ `sheet_name` เพื่อเลือกอ่าน sheet ที่ต้องการจาก Excel ได้ 
15. อ่านหลาย sheet จากไฟล์ Excel และเข้าใจผลลัพธ์ที่เป็น dictionary ของ DataFrame ได้

## ลำดับเนื้อหา 
บทเรียนนี้ประกอบด้วยหัวข้อต่อไปนี้ 
1. เตรียม pandas สำหรับการนำเข้าไฟล์ 
2. รู้จักไฟล์ตัวอย่างในบทเรียนนี้ 
3. การอ่านไฟล์ CSV ภาษาไทยด้วย `pd.read_csv()` 
4. การใช้ `encoding` กับไฟล์ CSV ภาษาไทย 
5. การอ่านเพียงบางส่วนด้วย `nrows` 
6. การอ่านเฉพาะบาง column ด้วย `usecols` 
7. การกำหนดชนิดข้อมูลด้วย `dtype` 
8. การอ่านข้อมูลวันที่ด้วย `parse_dates` 
9. การอ่านไฟล์ Excel ด้วย `pd.read_excel()` 
10. การเลือก sheet ด้วย `sheet_name` 
11. การอ่านหลาย sheet จาก Excel 
12. การใช้ loop อ่านหลาย sheet 
13. การอ่านไฟล์จาก URL 
14. ไฟล์ text ที่มีตัวคั่นข้อมูล 
15. ทดลองอ่านข้อมูลจาก URL 
16. ทำความเข้าใจโครงสร้างข้อมูลประชากรรายอายุ
17. สร้างชื่อ column สำหรับข้อมูลรายอายุด้วย loop 
18. อ่านข้อมูลประชากรจาก URL พร้อมกำหนดชื่อ column
19. แบบฝึกหัดท้ายบท

## 1. เตรียม pandas สำหรับการนำเข้าไฟล์ 
ก่อนใช้คำสั่งอ่านไฟล์ ต้อง import pandas ก่อน 

โดยทั่วไปนิยมใช้ชื่อย่อว่า `pd`

In [1]:
import pandas as pd

## 2. รู้จักไฟล์ตัวอย่างในบทเรียนนี้ 

ในบทนี้เราจะใช้ข้อมูลต่อไปนี้ในการเรียน 

- ไฟล์ csv

```python 
msdhs_ops_mso_logbook.csv
```

- ไฟล์ excel

```python 
moac_opsmoac_fragile_farmer.xlsx
```

- ไฟล์ text จาก url

```python 
https://stat.bora.dopa.go.th/new_stat/file/6812/6812cc10.txt
```

โดยเราจะเก็บชื่อไฟล์หรือ path ไว้ในตัวแปร ซึ่งมีข้อดีคือ 
- แก้ชื่อไฟล์ได้ง่าย 
- ลดการพิมพ์ซ้ำ 
- ทำให้ code อ่านง่ายขึ้น 
- เหมาะกับการใช้ไฟล์เดิมหลายครั้ง แต่เปลี่ยน parameter ในการอ่านไฟล์

In [2]:
csv_file = "msdhs_ops_mso_logbook.csv"

excel_file = "moac_opsmoac_fragile_farmer.xlsx"

text_file_url = "https://stat.bora.dopa.go.th/new_stat/file/6812/6812cc10.txt"

## 3. การอ่านไฟล์ CSV ด้วย `pd.read_csv()`

คำสั่งหลักสำหรับอ่านไฟล์ CSV คือ

```python
pd.read_csv()
```

รูปแบบพื้นฐานคือ

```python
df = pd.read_csv("file_name.csv")
```

In [3]:
mso_df = pd.read_csv(csv_file, encoding="utf-8-sig") 
mso_df

,วันที่แก้ไขข้อมูลล่าสุด,ชื่อสสว.,รหัสครัวเรือน,รหัสประจำบ้าน,วันที่สร้างครัวเรือน,ลักษณะที่อยู่อาศัย,เลขบัตรประชาชน,เลขที่สมาชิกครัวเรือน,คำนำหน้า,ชื่อ,...,ปัญหาด้านการเป็นผู้เสียหายจากการค้ามนุษย์,ปัญหาด้านการเข้าไม่ถึงสิทธิและความเป็นธรรมในสังคม,ปัญหาด้านสภาพปัญหาสังคมอื่นๆ,ประเภทกลุ่มเป้าหมาย,รหัส cm,ชื่อ นามสกุล cm,หน่วยงานของ cm,ละติจูด,ลองจิจูด,สถานะการมีชีวิต
0,20230619 13:40:56.585,5,648ff7d6b64d06b6b0dc7f02,NaN,20230619 13:38:14.881,ที่ดินมีกรรมสิทธิ์ของตนเอง,5962222985aa759afbc332b96975a4c4d74a5f82a1ab8a...,648ff83350573a1fb581f155,เด็กหญิง,****,...,NaN,NaN,NaN,เด็กเล็ก,cm410011,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
1,20220717 18:56:10.147,3,62d3f7a6d6f101550054198b,7.004022e+10,20220717 18:51:02.889,ที่ดินของผู้อื่น,62252665a08c86140a798453916f866d01ff17b8683341...,62d3f8aa24e6c5bf7e6edfdf,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm700018,****,ศูนย์คุ้มครองคนไร้ที่พึ่งราชบุรี,****,****,1.0
2,20240125 13:34:46.941,5,62a05c9df5674ecdd3805787,NaN,20220806 15:23:57.850,ที่ดินมีกรรมสิทธิ์ของตนเอง,761904ae1beb77d5de38fcbf34e757020c89291e467225...,62a05c9d63390df4b203ffdc,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm430010,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
3,20240322 14:12:39.290,6,62908e378fa67bf5ad7d5555,NaN,20220527 15:39:19.197,ที่ดินมีกรรมสิทธิ์ของตนเอง,a94f022ed6a769f28d8cd2eefafed33d4185008de20a0b...,62908e3763390df4b20378f6,เด็กหญิง,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm340009,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
4,20230315 09:27:11.763,4,6319b26ca3a37508384e36fc,NaN,20220809 16:14:20.652,ที่ดินมีกรรมสิทธิ์ของตนเอง,704921c77c739b12a7551cadae9998e06e50848647c058...,64112cc9dafc21d42cd4cfac,เด็กชาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm300020,****,ศูนย์บริการคนพิการ จังหวัดนครราชสีมา,****,****,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,20250729 14:33:32.178,4,688878eaa933a405b21cc136,NaN,20250729 14:31:54.610,ที่ดินมีกรรมสิทธิ์ของตนเอง,67d95a0afee5ffbf4a2aef3dbe7827865b72bd2752947f...,688878ea10a681cf016c3af2,นาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm360060,****,ศูนย์คุ้มครองคนไร้ที่พึ่ง จ.ชัยภูมิ,****,****,NaN
96,20230317 19:08:15.489,5,628b008c8fa67bf5ad7d221b,NaN,20220523 10:33:32.270,ที่ดินมีกรรมสิทธิ์ของตนเอง,90ce58571dcc07099fa5afca0762ba63e7fb5315cc86b5...,628b008c63390df4b2031ef9,นาง,****,...,NaN,NaN,NaN,ผู้สูงอายุ,cm440005,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,NaN
97,20230501 09:19:26.395,2,63b6325b2617b2b7aa04dc0b,2.512025e+10,20230501 09:13:47.520,ที่ดินมีกรรมสิทธิ์ของตนเอง,836047cbd5cf2ee78691f2fa34e90ef263c19477c909ec...,63b6325ba27dc8bc520df3fe,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm270001,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
98,20250104 14:05:25.955,5,63e1d0d92617b2b7aa06096d,4.016010e+10,20230702 11:17:29.369,ที่ดินมีกรรมสิทธิ์ของตนเอง,30b9e4c9830d3770dc3dc2104f85713df60643a470f494...,63e1d0d95b95aa0f8b7846b8,นาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,ผู้สูงอายุ,cm400016,****,ศูนย์เรียนรู้การพัฒนาสตรีและครอบครัวรัตนาภา จั...,****,****,1.0


ไฟล์นี้เป็นข้อมูลระบบสมุดพกครอบครัวอิเล็กทรอนิกส์ (MSO-Logbook) ซึ่งเป็นข้อมูลระดับครัวเรือนและบุคคล

ที่เกี่ยวข้องกับงานด้านสังคมสงเคราะห์และการพัฒนาคุณภาพชีวิต ประกอบด้วยข้อมูลหลายมิติ 

ชุดข้อมูลนี้เหมาะสำหรับใช้สอนการนำเข้าไฟล์ด้วย `pandas` ผ่านคำสั่ง `pd.read_csv()` 

เพราะเป็นไฟล์ csv ที่มีโครงสร้างแบบตารางชัดเจน มีชื่อคอลัมน์จำนวนมาก 

และมีชนิดข้อมูลหลากหลาย ทั้งข้อความ ตัวเลข วันที่ รหัสประจำรายการ และค่าว่าง 

ซึ่งช่วยให้ผู้เรียนเห็นตัวอย่างข้อมูลจริงที่พบได้ในการทำงานด้านสังคมสงเคราะห์

## 4. การใช้ `encoding` กับไฟล์ csv ภาษาไทย 
`Encoding` คือวิธีที่คอมพิวเตอร์ใช้แปลงตัวอักษรให้เป็นรหัสตัวเลข เพื่อให้สามารถบันทึก อ่าน และแลกเปลี่ยนข้อมูลในไฟล์ได้ 

เช่น ตัวอักษรภาษาไทย ภาษาอังกฤษ ตัวเลข หรือสัญลักษณ์ต่าง ๆ ล้วนต้องถูกเข้ารหัสด้วยรูปแบบ `encoding` บางชนิด

การอ่านไฟล์ csv ต้องรู้ว่าไฟล์นั้นใช้ `encoding` แบบใด หากกำหนด `encoding` 

ไม่ตรงกับไฟล์จริง อาจทำให้เกิดปัญหา เช่น อ่านไฟล์ไม่ได้ หรือแสดงผลภาษาไทยเป็นตัวอักษรเพี้ยน อ่านไม่ออก

ไฟล์ csv ภาษาไทยอาจต้องกำหนด `encoding` ให้เหมาะสม 

`encoding` ที่มักพบบ่อย ได้แก่ 
| encoding | กรณีที่มักพบ | 
|---|---| 
| `utf-8` | ไฟล์มาตรฐานทั่วไป | 
| `utf-8-sig` | csv ที่ export จาก Excel หรือระบบบางประเภท | 
| `cp874` | ไฟล์ภาษาไทยจาก Windows หรือระบบเดิม | 
| `tis-620` | ไฟล์ภาษาไทยจากระบบราชการหรือระบบ legacy บางประเภท | 

ดังนั้น หากอ่านไฟล์ csv แล้วพบว่าภาษาไทยเพี้ยนหรือเกิด error ควรลองตรวจสอบและกำหนดค่า `encoding` ให้ตรงกับรูปแบบของไฟล์นั้น

## 5. การอ่านไฟล์เพียงบางส่วนด้วย `nrows` 
หากไฟล์มีขนาดใหญ่ เราอาจยังไม่ต้องอ่านทั้งไฟล์ในครั้งแรก 

สามารถใช้ `nrows` เพื่ออ่านเฉพาะจำนวนแถวที่ต้องการได้

In [4]:
mso_preview = pd.read_csv( 
    csv_file, 
    encoding="utf-8-sig", 
    nrows=5 
) 
mso_preview

,วันที่แก้ไขข้อมูลล่าสุด,ชื่อสสว.,รหัสครัวเรือน,รหัสประจำบ้าน,วันที่สร้างครัวเรือน,ลักษณะที่อยู่อาศัย,เลขบัตรประชาชน,เลขที่สมาชิกครัวเรือน,คำนำหน้า,ชื่อ,...,ปัญหาด้านการเป็นผู้เสียหายจากการค้ามนุษย์,ปัญหาด้านการเข้าไม่ถึงสิทธิและความเป็นธรรมในสังคม,ปัญหาด้านสภาพปัญหาสังคมอื่นๆ,ประเภทกลุ่มเป้าหมาย,รหัส cm,ชื่อ นามสกุล cm,หน่วยงานของ cm,ละติจูด,ลองจิจูด,สถานะการมีชีวิต
0,20230619 13:40:56.585,5,648ff7d6b64d06b6b0dc7f02,NaN,20230619 13:38:14.881,ที่ดินมีกรรมสิทธิ์ของตนเอง,5962222985aa759afbc332b96975a4c4d74a5f82a1ab8a...,648ff83350573a1fb581f155,เด็กหญิง,****,...,NaN,NaN,NaN,เด็กเล็ก,cm410011,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1
1,20220717 18:56:10.147,3,62d3f7a6d6f101550054198b,7.004022e+10,20220717 18:51:02.889,ที่ดินของผู้อื่น,62252665a08c86140a798453916f866d01ff17b8683341...,62d3f8aa24e6c5bf7e6edfdf,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm700018,****,ศูนย์คุ้มครองคนไร้ที่พึ่งราชบุรี,****,****,1
2,20240125 13:34:46.941,5,62a05c9df5674ecdd3805787,NaN,20220806 15:23:57.850,ที่ดินมีกรรมสิทธิ์ของตนเอง,761904ae1beb77d5de38fcbf34e757020c89291e467225...,62a05c9d63390df4b203ffdc,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm430010,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1
3,20240322 14:12:39.290,6,62908e378fa67bf5ad7d5555,NaN,20220527 15:39:19.197,ที่ดินมีกรรมสิทธิ์ของตนเอง,a94f022ed6a769f28d8cd2eefafed33d4185008de20a0b...,62908e3763390df4b20378f6,เด็กหญิง,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm340009,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1
4,20230315 09:27:11.763,4,6319b26ca3a37508384e36fc,NaN,20220809 16:14:20.652,ที่ดินมีกรรมสิทธิ์ของตนเอง,704921c77c739b12a7551cadae9998e06e50848647c058...,64112cc9dafc21d42cd4cfac,เด็กชาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm300020,****,ศูนย์บริการคนพิการ จังหวัดนครราชสีมา,****,****,1


## 6. การเลือกอ่านเฉพาะบาง column ด้วย `usecols` 
ไฟล์จริงมักมี column จำนวนมาก แต่เราอาจต้องใช้เพียงบาง column 

เราสามารถใช้ `usecols` เพื่อเลือกอ่านเฉพาะ column ที่จำเป็นได้

In [5]:
mso_usecols = [ 
    "รหัสครัวเรือน", 
    "วันที่สร้างครัวเรือน", 
    "จังหวัด", 
    "อายุ", 
    "เพศ", 
    "ระดับการศึกษา", 
    "อาชีพหลัก", 
    "ประเภทกลุ่มเป้าหมาย", 
    "หน่วยงานของ cm" ] 

mso_selected_df = pd.read_csv( 
    csv_file, 
    encoding="utf-8-sig", 
    usecols=mso_usecols 
) 

mso_selected_df

,รหัสครัวเรือน,วันที่สร้างครัวเรือน,อายุ,เพศ,จังหวัด,ระดับการศึกษา,อาชีพหลัก,ประเภทกลุ่มเป้าหมาย,หน่วยงานของ cm
0,648ff7d6b64d06b6b0dc7f02,20230619 13:38:14.881,3,หญิง,อุดรธานี,ไม่ได้เรียนหนังสือ,เกษตรกรรม (พืช ปศุสัตว์ ประมง),เด็กเล็ก,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...
1,62d3f7a6d6f101550054198b,20220717 18:51:02.889,55,หญิง,ราชบุรี,ไม่ได้เรียนหนังสือ,NaN,วัยผู้ใหญ่/วัยแรงงาน,ศูนย์คุ้มครองคนไร้ที่พึ่งราชบุรี
2,62a05c9df5674ecdd3805787,20220806 15:23:57.850,33,หญิง,หนองคาย,มัธยมศึกษาตอนต้น,NaN,วัยผู้ใหญ่/วัยแรงงาน,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...
3,62908e378fa67bf5ad7d5555,20220527 15:39:19.197,5,หญิง,อุบลราชธานี,ไม่ได้เรียนหนังสือ,NaN,เด็กเล็ก,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...
4,6319b26ca3a37508384e36fc,20220809 16:14:20.652,3,ชาย,นครราชสีมา,ไม่ได้เรียนหนังสือ,NaN,เด็กเล็ก,ศูนย์บริการคนพิการ จังหวัดนครราชสีมา
...,...,...,...,...,...,...,...,...,...
95,688878eaa933a405b21cc136,20250729 14:31:54.610,59,NaN,ชัยภูมิ,NaN,เกษตรกรรม (พืช ปศุสัตว์ ประมง),วัยผู้ใหญ่/วัยแรงงาน,ศูนย์คุ้มครองคนไร้ที่พึ่ง จ.ชัยภูมิ
96,628b008c8fa67bf5ad7d221b,20220523 10:33:32.270,79,NaN,มหาสารคาม,NaN,NaN,ผู้สูงอายุ,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...
97,63b6325b2617b2b7aa04dc0b,20230501 09:13:47.520,51,หญิง,สระแก้ว,ประถมศึกษา,รับจ้างทั่วไป,วัยผู้ใหญ่/วัยแรงงาน,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...
98,63e1d0d92617b2b7aa06096d,20230702 11:17:29.369,81,ชาย,ขอนแก่น,ไม่ได้เรียนหนังสือ,NaN,ผู้สูงอายุ,ศูนย์เรียนรู้การพัฒนาสตรีและครอบครัวรัตนาภา จั...


## 7. การกำหนดชนิดข้อมูลด้วย `dtype` 
บาง column แม้จะดูเหมือนตัวเลข แต่ไม่ควรอ่านเป็นตัวเลข 

ตัวอย่างเช่น column ที่เป็นรหัส 
- `รหัสครัวเรือน` 
- `รหัสประจำบ้าน` 
- `รหัส cm` 

ข้อมูลเหล่านี้เป็น “รหัส” ไม่ใช่ตัวเลขสำหรับคำนวณ 

ถ้า pandas อ่านเป็นตัวเลข อาจเกิดปัญหา เช่น 
- เลข 0 ด้านหน้าหาย 
- ตัวเลขยาวถูกแสดงเป็น scientific notation 
- รหัสถูกตีความผิดว่าเป็นค่าที่นำไปบวก ลบ คูณ หารได้ 

ดังนั้นควรกำหนด `dtype` ให้ column ประเภทนี้เป็น string

In [6]:
mso_dtype = { 
    "รหัสครัวเรือน": "string", 
    "รหัสประจำบ้าน": "string", 
    "รหัส cm": "string" 
} 

mso_df = pd.read_csv( 
    csv_file, 
    encoding="utf-8-sig", 
    dtype=mso_dtype 
) 

mso_df

,วันที่แก้ไขข้อมูลล่าสุด,ชื่อสสว.,รหัสครัวเรือน,รหัสประจำบ้าน,วันที่สร้างครัวเรือน,ลักษณะที่อยู่อาศัย,เลขบัตรประชาชน,เลขที่สมาชิกครัวเรือน,คำนำหน้า,ชื่อ,...,ปัญหาด้านการเป็นผู้เสียหายจากการค้ามนุษย์,ปัญหาด้านการเข้าไม่ถึงสิทธิและความเป็นธรรมในสังคม,ปัญหาด้านสภาพปัญหาสังคมอื่นๆ,ประเภทกลุ่มเป้าหมาย,รหัส cm,ชื่อ นามสกุล cm,หน่วยงานของ cm,ละติจูด,ลองจิจูด,สถานะการมีชีวิต
0,20230619 13:40:56.585,5,648ff7d6b64d06b6b0dc7f02,<NA>,20230619 13:38:14.881,ที่ดินมีกรรมสิทธิ์ของตนเอง,5962222985aa759afbc332b96975a4c4d74a5f82a1ab8a...,648ff83350573a1fb581f155,เด็กหญิง,****,...,NaN,NaN,NaN,เด็กเล็ก,cm410011,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
1,20220717 18:56:10.147,3,62d3f7a6d6f101550054198b,70040223124,20220717 18:51:02.889,ที่ดินของผู้อื่น,62252665a08c86140a798453916f866d01ff17b8683341...,62d3f8aa24e6c5bf7e6edfdf,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm700018,****,ศูนย์คุ้มครองคนไร้ที่พึ่งราชบุรี,****,****,1.0
2,20240125 13:34:46.941,5,62a05c9df5674ecdd3805787,<NA>,20220806 15:23:57.850,ที่ดินมีกรรมสิทธิ์ของตนเอง,761904ae1beb77d5de38fcbf34e757020c89291e467225...,62a05c9d63390df4b203ffdc,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm430010,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
3,20240322 14:12:39.290,6,62908e378fa67bf5ad7d5555,<NA>,20220527 15:39:19.197,ที่ดินมีกรรมสิทธิ์ของตนเอง,a94f022ed6a769f28d8cd2eefafed33d4185008de20a0b...,62908e3763390df4b20378f6,เด็กหญิง,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm340009,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
4,20230315 09:27:11.763,4,6319b26ca3a37508384e36fc,<NA>,20220809 16:14:20.652,ที่ดินมีกรรมสิทธิ์ของตนเอง,704921c77c739b12a7551cadae9998e06e50848647c058...,64112cc9dafc21d42cd4cfac,เด็กชาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm300020,****,ศูนย์บริการคนพิการ จังหวัดนครราชสีมา,****,****,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,20250729 14:33:32.178,4,688878eaa933a405b21cc136,<NA>,20250729 14:31:54.610,ที่ดินมีกรรมสิทธิ์ของตนเอง,67d95a0afee5ffbf4a2aef3dbe7827865b72bd2752947f...,688878ea10a681cf016c3af2,นาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm360060,****,ศูนย์คุ้มครองคนไร้ที่พึ่ง จ.ชัยภูมิ,****,****,NaN
96,20230317 19:08:15.489,5,628b008c8fa67bf5ad7d221b,<NA>,20220523 10:33:32.270,ที่ดินมีกรรมสิทธิ์ของตนเอง,90ce58571dcc07099fa5afca0762ba63e7fb5315cc86b5...,628b008c63390df4b2031ef9,นาง,****,...,NaN,NaN,NaN,ผู้สูงอายุ,cm440005,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,NaN
97,20230501 09:19:26.395,2,63b6325b2617b2b7aa04dc0b,25120251099,20230501 09:13:47.520,ที่ดินมีกรรมสิทธิ์ของตนเอง,836047cbd5cf2ee78691f2fa34e90ef263c19477c909ec...,63b6325ba27dc8bc520df3fe,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm270001,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
98,20250104 14:05:25.955,5,63e1d0d92617b2b7aa06096d,40160104483,20230702 11:17:29.369,ที่ดินมีกรรมสิทธิ์ของตนเอง,30b9e4c9830d3770dc3dc2104f85713df60643a470f494...,63e1d0d95b95aa0f8b7846b8,นาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,ผู้สูงอายุ,cm400016,****,ศูนย์เรียนรู้การพัฒนาสตรีและครอบครัวรัตนาภา จั...,****,****,1.0


## 8. การอ่านวันที่ด้วย `parse_dates` 
บางไฟล์มี column วันที่ เช่น 
- `วันที่แก้ไขข้อมูลล่าสุด` 
- `วันที่สร้างครัวเรือน` 

เราสามารถใช้ `parse_dates` เพื่อให้ pandas พยายามอ่าน column เหล่านี้เป็นวันที่

In [7]:
mso_df = pd.read_csv( 
    csv_file, 
    encoding="utf-8-sig", 
    dtype=mso_dtype, 
    parse_dates=["วันที่แก้ไขข้อมูลล่าสุด", "วันที่สร้างครัวเรือน"] 
) 

mso_df

,วันที่แก้ไขข้อมูลล่าสุด,ชื่อสสว.,รหัสครัวเรือน,รหัสประจำบ้าน,วันที่สร้างครัวเรือน,ลักษณะที่อยู่อาศัย,เลขบัตรประชาชน,เลขที่สมาชิกครัวเรือน,คำนำหน้า,ชื่อ,...,ปัญหาด้านการเป็นผู้เสียหายจากการค้ามนุษย์,ปัญหาด้านการเข้าไม่ถึงสิทธิและความเป็นธรรมในสังคม,ปัญหาด้านสภาพปัญหาสังคมอื่นๆ,ประเภทกลุ่มเป้าหมาย,รหัส cm,ชื่อ นามสกุล cm,หน่วยงานของ cm,ละติจูด,ลองจิจูด,สถานะการมีชีวิต
0,2023-06-19 13:40:56.585,5,648ff7d6b64d06b6b0dc7f02,<NA>,2023-06-19 13:38:14.881,ที่ดินมีกรรมสิทธิ์ของตนเอง,5962222985aa759afbc332b96975a4c4d74a5f82a1ab8a...,648ff83350573a1fb581f155,เด็กหญิง,****,...,NaN,NaN,NaN,เด็กเล็ก,cm410011,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
1,2022-07-17 18:56:10.147,3,62d3f7a6d6f101550054198b,70040223124,2022-07-17 18:51:02.889,ที่ดินของผู้อื่น,62252665a08c86140a798453916f866d01ff17b8683341...,62d3f8aa24e6c5bf7e6edfdf,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm700018,****,ศูนย์คุ้มครองคนไร้ที่พึ่งราชบุรี,****,****,1.0
2,2024-01-25 13:34:46.941,5,62a05c9df5674ecdd3805787,<NA>,2022-08-06 15:23:57.850,ที่ดินมีกรรมสิทธิ์ของตนเอง,761904ae1beb77d5de38fcbf34e757020c89291e467225...,62a05c9d63390df4b203ffdc,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm430010,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
3,2024-03-22 14:12:39.290,6,62908e378fa67bf5ad7d5555,<NA>,2022-05-27 15:39:19.197,ที่ดินมีกรรมสิทธิ์ของตนเอง,a94f022ed6a769f28d8cd2eefafed33d4185008de20a0b...,62908e3763390df4b20378f6,เด็กหญิง,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm340009,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
4,2023-03-15 09:27:11.763,4,6319b26ca3a37508384e36fc,<NA>,2022-08-09 16:14:20.652,ที่ดินมีกรรมสิทธิ์ของตนเอง,704921c77c739b12a7551cadae9998e06e50848647c058...,64112cc9dafc21d42cd4cfac,เด็กชาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,เด็กเล็ก,cm300020,****,ศูนย์บริการคนพิการ จังหวัดนครราชสีมา,****,****,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,2025-07-29 14:33:32.178,4,688878eaa933a405b21cc136,<NA>,2025-07-29 14:31:54.610,ที่ดินมีกรรมสิทธิ์ของตนเอง,67d95a0afee5ffbf4a2aef3dbe7827865b72bd2752947f...,688878ea10a681cf016c3af2,นาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm360060,****,ศูนย์คุ้มครองคนไร้ที่พึ่ง จ.ชัยภูมิ,****,****,NaN
96,2023-03-17 19:08:15.489,5,628b008c8fa67bf5ad7d221b,<NA>,2022-05-23 10:33:32.270,ที่ดินมีกรรมสิทธิ์ของตนเอง,90ce58571dcc07099fa5afca0762ba63e7fb5315cc86b5...,628b008c63390df4b2031ef9,นาง,****,...,NaN,NaN,NaN,ผู้สูงอายุ,cm440005,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,NaN
97,2023-05-01 09:19:26.395,2,63b6325b2617b2b7aa04dc0b,25120251099,2023-05-01 09:13:47.520,ที่ดินมีกรรมสิทธิ์ของตนเอง,836047cbd5cf2ee78691f2fa34e90ef263c19477c909ec...,63b6325ba27dc8bc520df3fe,นางสาว,****,...,NaN,NaN,NaN,วัยผู้ใหญ่/วัยแรงงาน,cm270001,****,สำนักงานพัฒนาสังคมและความมั่นคงของมนุษย์จังหวั...,****,****,1.0
98,2025-01-04 14:05:25.955,5,63e1d0d92617b2b7aa06096d,40160104483,2023-07-02 11:17:29.369,ที่ดินมีกรรมสิทธิ์ของตนเอง,30b9e4c9830d3770dc3dc2104f85713df60643a470f494...,63e1d0d95b95aa0f8b7846b8,นาย,****,...,NaN,ไม่สามารถเข้าถึงบริการของรัฐ|,NaN,ผู้สูงอายุ,cm400016,****,ศูนย์เรียนรู้การพัฒนาสตรีและครอบครัวรัตนาภา จั...,****,****,1.0


## 9. การอ่านไฟล์ Excel ด้วย `pd.read_excel()`

คำสั่งหลักสำหรับอ่านไฟล์ Excel คือ

```python
pd.read_excel()
```

รูปแบบพื้นฐานคือ

```python
df = pd.read_excel("file_name.xlsx")
```

In [8]:
%pip install openpyxl


[notice] A new release of pip is available: 26.0.1 -> 26.1.2
[notice] To update, run: pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


ใน pandas เราสามารถอ่านไฟล์ Excel ได้ด้วยคำสั่ง `pd.read_excel()`

แต่ pandas ไม่ได้อ่านไฟล์ Excel `.xlsx` ด้วยตัวเองทั้งหมด 

จำเป็นต้องใช้ไลบรารีเสริมที่ทำหน้าที่เป็น “ตัวช่วยอ่านไฟล์ Excel” หรือที่เรียกว่า `engine`

สำหรับไฟล์ Excel นามสกุล `.xlsx` ไลบรารีที่นิยมใช้คือ `openpyxl`

In [9]:
farmer_df = pd.read_excel(excel_file) 
farmer_df

,department_code,department,pid,gender,birthdate,province_code,amphur_code,tambon_code,province,amphur,...,alro_allocated_area_ngan,alro_allocated_area_wa,is_alro,alro_postcode,updated_at,cpd_year_data,doae_reg_year,doae_corporation_be,doae_member_flag,doae_kaset_org_name
0,cpd,กรมส่งเสริมสหกรณ์,60e3f4907b85b690287f68a641b01b3d8392fbc28398f5...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
1,cpd,กรมส่งเสริมสหกรณ์,1ac1b0b4ad588a2c883621160bc0bbb5ca122dd42743b7...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
2,cpd,กรมส่งเสริมสหกรณ์,55e7b240314a3b5087b8ee4c376b52f229edda0d64761d...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
3,cpd,กรมส่งเสริมสหกรณ์,62a25e21b1ee1604c8cd4b812d81cf66854dd49cd25c6c...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
4,cpd,กรมส่งเสริมสหกรณ์,4eb76a041e7389e4ae9d22ecefd36a5ac999f4c5ea5ebe...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
5,cpd,กรมส่งเสริมสหกรณ์,fce9763eb646cdac1d4b6e1ce147595ab8cb1857263a1e...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
6,cpd,กรมส่งเสริมสหกรณ์,6673b22c71d3c48961157c0c240bce5531c57da907544b...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
7,cpd,กรมส่งเสริมสหกรณ์,82b98885daa5cd9368b364da0730d094aad99594ee3ce9...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
8,cpd,กรมส่งเสริมสหกรณ์,a177c5068a6c1aa29db77b7039a830e455900297f1b104...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN
9,cpd,กรมส่งเสริมสหกรณ์,6538056606ffebaa3730de51cfbe595b3230fa0c7a578f...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,2023,NaN,NaN,NaN,NaN


ไฟล์นี้เป็นข้อมูลเกษตรกรเปราะบางจากหลายหน่วยงานในกระทรวงเกษตรและสหกรณ์ 

ซึ่งข้อมูลของแต่ละหน่วยจะถูกเก็บไว้ในแต่ละ `worksheet` ได้แก่

- v_cpd_fragile
- v_dld_fragile
- v_doae_fragile
- v_dof_fragile
- v_raot_fragile

โดย default `pd.read_excel()` จะอ่าน sheet แรกของไฟล์

จากไฟล์ตัวอย่างนี้ sheet แรกคือ

```python
v_cpd_fragile
```

แต่ไฟล์ Excel นี้มีหลาย sheet ดังนั้นในการทำงานจริงควรกำหนด sheet_name ให้ชัดเจน

In [10]:
doae_df = pd.read_excel( 
    excel_file, 
    sheet_name="v_doae_fragile" ) 

doae_df

,department_code,department,pid,gender,birthdate,province_code,amphur_code,tambon_code,province,amphur,...,alro_allocated_area_ngan,alro_allocated_area_wa,is_alro,alro_postcode,updated_at,cpd_year_data,doae_reg_year,doae_corporation_be,doae_member_flag,doae_kaset_org_name
0,doae,กรมส่งเสริมการเกษตร,da98e3740c58278c3d0a4f01b23109a36b4a5a42a5fa72...,NaN,NaN,53.0,3.0,4.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2023.0,NaN,สมาชิก,อื่นๆ
1,doae,กรมส่งเสริมการเกษตร,5295ea9f23f585562a5af56f1f3a6d12c463140aa8b367...,NaN,NaN,95.0,8.0,4.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,NaN,สมาชิก,NaN
2,doae,กรมส่งเสริมการเกษตร,666e131e16b53c38ae8ee41917b2ab5ddd4f0e8674b19f...,NaN,NaN,93.0,7.0,3.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,1.0,หัวหน้าครัวเรือน,NaN
3,doae,กรมส่งเสริมการเกษตร,bbd58a88987d6f79a3733dfd4885f0b74028004379ded3...,NaN,NaN,86.0,5.0,2.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,NaN,สมาชิก,NaN
4,doae,กรมส่งเสริมการเกษตร,ebdb71900fee10f50d9f5b638e8a64d1fe3b7786fb69b7...,NaN,NaN,30.0,11.0,6.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2016.0,NaN,สมาชิก,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
65,doae,กรมส่งเสริมการเกษตร,b437149cc5b6a79f8866914fd84b644196ed99ba4f3d62...,NaN,NaN,96.0,6.0,7.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,1.0,หัวหน้าครัวเรือน,NaN
66,doae,กรมส่งเสริมการเกษตร,b1ce3ba83cfcfc5c21414fe2c496a37b635f56cb4de98c...,NaN,NaN,61.0,2.0,8.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2024.0,NaN,สมาชิก,อื่นๆ
67,doae,กรมส่งเสริมการเกษตร,5bdabb8ac705d3ba6fab31c9f7f1a48c37455ac548d72a...,NaN,NaN,53.0,7.0,10.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,1.0,หัวหน้าครัวเรือน,ธกส.
68,doae,กรมส่งเสริมการเกษตร,d066b479f17041cb1297e115b3ed0c8e8b724769855a97...,NaN,NaN,12.0,6.0,10.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2016.0,NaN,สมาชิก,อื่นๆ


## 10. การอ่านรายชื่อ sheet ทั้งหมดใน Excel 
ก่อนเลือกอ่าน sheet เราสามารถตรวจสอบได้ว่าไฟล์ Excel มี sheet อะไรบ้าง 

ใช้ `pd.ExcelFile()` เพื่ออ่านโครงสร้างของไฟล์ Excel

In [11]:
excel_data = pd.ExcelFile(excel_file) 

sheet_names = excel_data.sheet_names 
sheet_names

['v_cpd_fragile',
 'v_dld_fragile',
 'v_doae_fragile',
 'v_dof_fragile',
 'v_raot_fragile']

## 11. การอ่านหลาย sheet จาก Excel 
เราสามารถอ่านหลาย sheet จากไฟล์ Excel ได้โดยกำหนด `sheet_name` เป็น list

In [12]:
selected_sheets = [ 
    "v_cpd_fragile", 
    "v_dld_fragile", 
    "v_doae_fragile" 
] 

farmer_sheets = pd.read_excel( 
    excel_file, 
    sheet_name=selected_sheets 
) 

farmer_sheets

{'v_cpd_fragile':    department_code         department  \
 0              cpd  กรมส่งเสริมสหกรณ์   
 1              cpd  กรมส่งเสริมสหกรณ์   
 2              cpd  กรมส่งเสริมสหกรณ์   
 3              cpd  กรมส่งเสริมสหกรณ์   
 4              cpd  กรมส่งเสริมสหกรณ์   
 5              cpd  กรมส่งเสริมสหกรณ์   
 6              cpd  กรมส่งเสริมสหกรณ์   
 7              cpd  กรมส่งเสริมสหกรณ์   
 8              cpd  กรมส่งเสริมสหกรณ์   
 9              cpd  กรมส่งเสริมสหกรณ์   
 10             cpd  กรมส่งเสริมสหกรณ์   
 11             cpd  กรมส่งเสริมสหกรณ์   
 12             cpd  กรมส่งเสริมสหกรณ์   
 13             cpd  กรมส่งเสริมสหกรณ์   
 14             cpd  กรมส่งเสริมสหกรณ์   
 
                                                   pid  gender  birthdate  \
 0   60e3f4907b85b690287f68a641b01b3d8392fbc28398f5...     NaN        NaN   
 1   1ac1b0b4ad588a2c883621160bc0bbb5ca122dd42743b7...     NaN        NaN   
 2   55e7b240314a3b5087b8ee4c376b52f229edda0d64761d...     NaN        NaN   
 

เมื่ออ่านหลาย sheet พร้อมกัน ผลลัพธ์ที่ได้จะเป็น dictionary โดย 
- key คือชื่อ sheet 
- value คือ DataFrame ของ sheet นั้น 

เราสามารถเลือก DataFrame ของแต่ละ sheet ได้ด้วย key

In [13]:
farmer_sheets["v_doae_fragile"]

,department_code,department,pid,gender,birthdate,province_code,amphur_code,tambon_code,province,amphur,...,alro_allocated_area_ngan,alro_allocated_area_wa,is_alro,alro_postcode,updated_at,cpd_year_data,doae_reg_year,doae_corporation_be,doae_member_flag,doae_kaset_org_name
0,doae,กรมส่งเสริมการเกษตร,da98e3740c58278c3d0a4f01b23109a36b4a5a42a5fa72...,NaN,NaN,53.0,3.0,4.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2023.0,NaN,สมาชิก,อื่นๆ
1,doae,กรมส่งเสริมการเกษตร,5295ea9f23f585562a5af56f1f3a6d12c463140aa8b367...,NaN,NaN,95.0,8.0,4.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,NaN,สมาชิก,NaN
2,doae,กรมส่งเสริมการเกษตร,666e131e16b53c38ae8ee41917b2ab5ddd4f0e8674b19f...,NaN,NaN,93.0,7.0,3.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,1.0,หัวหน้าครัวเรือน,NaN
3,doae,กรมส่งเสริมการเกษตร,bbd58a88987d6f79a3733dfd4885f0b74028004379ded3...,NaN,NaN,86.0,5.0,2.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,NaN,สมาชิก,NaN
4,doae,กรมส่งเสริมการเกษตร,ebdb71900fee10f50d9f5b638e8a64d1fe3b7786fb69b7...,NaN,NaN,30.0,11.0,6.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2016.0,NaN,สมาชิก,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
65,doae,กรมส่งเสริมการเกษตร,b437149cc5b6a79f8866914fd84b644196ed99ba4f3d62...,NaN,NaN,96.0,6.0,7.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,1.0,หัวหน้าครัวเรือน,NaN
66,doae,กรมส่งเสริมการเกษตร,b1ce3ba83cfcfc5c21414fe2c496a37b635f56cb4de98c...,NaN,NaN,61.0,2.0,8.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2024.0,NaN,สมาชิก,อื่นๆ
67,doae,กรมส่งเสริมการเกษตร,5bdabb8ac705d3ba6fab31c9f7f1a48c37455ac548d72a...,NaN,NaN,53.0,7.0,10.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2015.0,1.0,หัวหน้าครัวเรือน,ธกส.
68,doae,กรมส่งเสริมการเกษตร,d066b479f17041cb1297e115b3ed0c8e8b724769855a97...,NaN,NaN,12.0,6.0,10.0,NaN,NaN,...,NaN,NaN,NaN,NaN,20250825,NaN,2016.0,NaN,สมาชิก,อื่นๆ


## 12. การใช้ loop เพื่ออ่านหลาย sheet 
เราสามารถประยุกต์ใช้ loop เพื่ออ่านหลาย sheet จาก Excel ได้เช่นกัน

In [14]:
all_farmer_sheets = {} 

for sheet in sheet_names: 
    all_farmer_sheets[sheet] = pd.read_excel( 
        excel_file, 
        sheet_name=sheet 
    ) 

all_farmer_sheets

{'v_cpd_fragile':    department_code         department  \
 0              cpd  กรมส่งเสริมสหกรณ์   
 1              cpd  กรมส่งเสริมสหกรณ์   
 2              cpd  กรมส่งเสริมสหกรณ์   
 3              cpd  กรมส่งเสริมสหกรณ์   
 4              cpd  กรมส่งเสริมสหกรณ์   
 5              cpd  กรมส่งเสริมสหกรณ์   
 6              cpd  กรมส่งเสริมสหกรณ์   
 7              cpd  กรมส่งเสริมสหกรณ์   
 8              cpd  กรมส่งเสริมสหกรณ์   
 9              cpd  กรมส่งเสริมสหกรณ์   
 10             cpd  กรมส่งเสริมสหกรณ์   
 11             cpd  กรมส่งเสริมสหกรณ์   
 12             cpd  กรมส่งเสริมสหกรณ์   
 13             cpd  กรมส่งเสริมสหกรณ์   
 14             cpd  กรมส่งเสริมสหกรณ์   
 
                                                   pid  gender  birthdate  \
 0   60e3f4907b85b690287f68a641b01b3d8392fbc28398f5...     NaN        NaN   
 1   1ac1b0b4ad588a2c883621160bc0bbb5ca122dd42743b7...     NaN        NaN   
 2   55e7b240314a3b5087b8ee4c376b52f229edda0d64761d...     NaN        NaN   
 

เราสามารถใช้ `usecols` เพื่อกำหนดให้นำเข้าเฉพาะ column และ 

`dtype` เพื่อกำหนดประเภทของ column กับ `pd.read_excel()`

ได้เช่นเดียวกับ `pd.read_csv()`

In [15]:
farmer_usecols = [ 
    "department_code", 
    "department", 
    "pid", 
    "province_code", 
    "province",
    "amphur_code",
    "amphur",
    "tambon_code", 
    "tambon"
]

farmer_dtype = { 
    "pid": "string", 
    "province_code": "string", 
    "amphur_code": "string", 
    "tambon_code": "string" 
}

doae_selected_df = pd.read_excel( 
    excel_file, 
    sheet_name="v_doae_fragile", 
    usecols=farmer_usecols,
    dtype=farmer_dtype 
) 

doae_selected_df

,department_code,department,pid,province_code,amphur_code,tambon_code,province,amphur,tambon
0,doae,กรมส่งเสริมการเกษตร,da98e3740c58278c3d0a4f01b23109a36b4a5a42a5fa72...,53,03,04,NaN,NaN,NaN
1,doae,กรมส่งเสริมการเกษตร,5295ea9f23f585562a5af56f1f3a6d12c463140aa8b367...,95,08,04,NaN,NaN,NaN
2,doae,กรมส่งเสริมการเกษตร,666e131e16b53c38ae8ee41917b2ab5ddd4f0e8674b19f...,93,07,03,NaN,NaN,NaN
3,doae,กรมส่งเสริมการเกษตร,bbd58a88987d6f79a3733dfd4885f0b74028004379ded3...,86,05,02,NaN,NaN,NaN
4,doae,กรมส่งเสริมการเกษตร,ebdb71900fee10f50d9f5b638e8a64d1fe3b7786fb69b7...,30,11,06,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...
65,doae,กรมส่งเสริมการเกษตร,b437149cc5b6a79f8866914fd84b644196ed99ba4f3d62...,96,06,07,NaN,NaN,NaN
66,doae,กรมส่งเสริมการเกษตร,b1ce3ba83cfcfc5c21414fe2c496a37b635f56cb4de98c...,61,02,08,NaN,NaN,NaN
67,doae,กรมส่งเสริมการเกษตร,5bdabb8ac705d3ba6fab31c9f7f1a48c37455ac548d72a...,53,07,10,NaN,NaN,NaN
68,doae,กรมส่งเสริมการเกษตร,d066b479f17041cb1297e115b3ed0c8e8b724769855a97...,12,06,10,NaN,NaN,NaN


## 13. การอ่านไฟล์จาก URL 
นอกจากการอ่านไฟล์ที่อยู่ในเครื่องแล้ว pandas ยังสามารถอ่านไฟล์จาก URL ได้โดยตรง 

ถ้า URL ชี้ไปยังไฟล์ข้อมูล เช่น CSV หรือ text file เราสามารถส่ง URL เข้าไปใน `pd.read_csv()` ได้ 

ตัวอย่างในบทนี้คือไฟล์ข้อมูลจำนวนประชากรแยกรายอายุของจังหวัดกรุงเทพมหานคร ปี 2568 

```python 
text_file_url = "https://stat.bora.dopa.go.th/new_stat/file/6812/6812cc10.txt"
```

ไฟล์นี้เป็นไฟล์ `.txt` แต่ยังสามารถใช้ `pd.read_csv()` อ่านได้ เพราะข้อมูลในไฟล์มีตัวคั่นระหว่าง field



## 14. ไฟล์ text ที่มีตัวคั่นข้อมูล 
ไฟล์ `.txt` ไม่ได้แปลว่าอ่านด้วย pandas ไม่ได้ 

ถ้าไฟล์ text มีโครงสร้างเป็นตาราง และมีตัวคั่นข้อมูลชัดเจน เช่น 
- comma `,` 
- semicolon `;` 
- tab `\t` 
- pipe `|` 

เราสามารถใช้ `pd.read_csv()` อ่านไฟล์นั้นได้ โดยกำหนด `sep` 

สำหรับไฟล์ประชากรจากกรมการปกครอง ตัวคั่นข้อมูลคือ pipe หรือ `|` 

ดังนั้นต้องใช้ 

```python 
sep="|"
```

## 15. ทดลองอ่านข้อมูลจาก URL 
ก่อนอ่านข้อมูลจริงทั้งหมด เราควรทดลองอ่านเพียงบางแถวด้วย `nrows`

In [16]:
population_preview = pd.read_csv( 
    text_file_url, 
    sep="|",
    nrows=5 
)

population_preview

,กรุงเทพมหานคร,11640,10838,14323,13274,15588,14635,15814,14567,17733,...,37804,28545,66349,1333,1150,2483,2526272,2896296,5422568,Unnamed: 220
0,เขตพระนคร,52,40,55,47,73,48,70,64,78,...,388,256,644,1,3,4,18723,19685,38408,NaN
1,แขวงพระบรมมหาราชวัง,4,2,7,2,3,4,2,5,4,...,17,16,33,1,0,1,1684,1104,2788,NaN
2,แขวงวังบูรพาภิรมย์,10,9,12,10,20,8,14,15,26,...,110,90,200,0,0,0,4172,4066,8238,NaN
3,แขวงวัดราชบพิธ,5,6,4,2,5,3,1,3,1,...,24,23,47,0,0,0,1218,1509,2727,NaN
4,แขวงสำราญราษฎร์,4,1,5,1,6,5,5,3,5,...,35,12,47,0,0,0,1332,1260,2592,NaN


จากผลลัพธ์เราจะสังเกตุเห็นว่าชื่อคอลัมน์ของข้อมูลดูผิดปกติ

เนื่องจากไฟล์นี้ไม่มีชื่อคอลัมน์หรือ `header`

`pandas` จึงเข้าใจว่าแถวแรกของไฟล์เป็นชื่อคอลัมน์

เราต้องกำหนด

```python 
header=None
```

เพื่อบอก pandas ว่าแถวแรกของไฟล์ไม่ใช่ชื่อคอลัมน์

In [17]:
population_preview = pd.read_csv( 
    text_file_url, 
    sep="|", 
    header=None, 
    nrows=5 
) 

population_preview

,0,1,2,3,4,5,6,7,8,9,...,211,212,213,214,215,216,217,218,219,220
0,กรุงเทพมหานคร,11640,10838,14323,13274,15588,14635,15814,14567,17733,...,37804,28545,66349,1333,1150,2483,2526272,2896296,5422568,NaN
1,เขตพระนคร,52,40,55,47,73,48,70,64,78,...,388,256,644,1,3,4,18723,19685,38408,NaN
2,แขวงพระบรมมหาราชวัง,4,2,7,2,3,4,2,5,4,...,17,16,33,1,0,1,1684,1104,2788,NaN
3,แขวงวังบูรพาภิรมย์,10,9,12,10,20,8,14,15,26,...,110,90,200,0,0,0,4172,4066,8238,NaN
4,แขวงวัดราชบพิธ,5,6,4,2,5,3,1,3,1,...,24,23,47,0,0,0,1218,1509,2727,NaN


จากผลลัพธ์จะเห็นว่า pandas อ่านข้อมูลได้ แต่ชื่อ column ยังเป็นตัวเลข เช่น 0, 1, 2, ... 

เพราะ `pandas` ไม่สามารถหา `header` ได้

ขั้นตอนถัดไปคือสร้างชื่อคอลัมน์เองจากโครงสร้างข้อมูลที่หน่วยงานอธิบายไว้

## 16. ทำความเข้าใจโครงสร้างข้อมูลประชากรรายอายุ 
จากโครงสร้างข้อมูลของกรมการปกครองตามลิงค์ต่อไปนี้

```text
https://stat.bora.dopa.go.th/new_stat/webPage/statByAge.php
```

ไฟล์จำนวนประชากรแยกรายอายุมี field หลัก ๆ ดังนี้

1. `DESC-CCAATT` รายละเอียดจังหวัด หรืออำเภอ หรือตำบล 
2. ชุดข้อมูลประชากรชายและหญิงรายอายุ ตั้งแต่อายุ 0 ถึง 101 โดยอายุ 0 หมายถึงอายุน้อยกว่า 1 ปี และอายุ 101 หมายถึงอายุมากกว่า 100 ปี 
3. ข้อมูลกลุ่มพิเศษและยอดรวม เช่น 
- ประชากรชาย/หญิงเกิดปีจันทรคติ 
- ประชากรชาย/หญิงที่มีชื่ออยู่ในทะเบียนบ้านกลาง 
- ประชากรชาย/หญิงที่มิใช่สัญชาติไทย 
- ประชากรชาย/หญิงที่อยู่ระหว่างการย้าย 
- ประชากรชายทั้งหมด 
- ประชากรหญิงทั้งหมด 
- ผลรวมประชากรทั้งหมด 

ดังนั้นเราจะต้องสร้างชื่อ column ให้ครบตามลำดับ field ของไฟล์

## 17. สร้างชื่อ column สำหรับข้อมูลรายอายุด้วย loop 

จากการประยุกต์ใช้เรื่อง `list` และ `loop` 

เราสามารถใช้ loop เพื่อสร้างชื่อ column รายอายุได้ แทนที่จะพิมพ์เองทั้งหมด 

ต้องการสร้าง column เช่น

```python 
male_age_0 
female_age_0 
male_age_1 
female_age_1 
... 
male_age_101
female_age_101
```

In [18]:
age_columns = [] 

for age in range(102): 
    age_columns.append(f"male_age_{age}") 
    age_columns.append(f"female_age_{age}") 
    
age_columns

['male_age_0',
 'female_age_0',
 'male_age_1',
 'female_age_1',
 'male_age_2',
 'female_age_2',
 'male_age_3',
 'female_age_3',
 'male_age_4',
 'female_age_4',
 'male_age_5',
 'female_age_5',
 'male_age_6',
 'female_age_6',
 'male_age_7',
 'female_age_7',
 'male_age_8',
 'female_age_8',
 'male_age_9',
 'female_age_9',
 'male_age_10',
 'female_age_10',
 'male_age_11',
 'female_age_11',
 'male_age_12',
 'female_age_12',
 'male_age_13',
 'female_age_13',
 'male_age_14',
 'female_age_14',
 'male_age_15',
 'female_age_15',
 'male_age_16',
 'female_age_16',
 'male_age_17',
 'female_age_17',
 'male_age_18',
 'female_age_18',
 'male_age_19',
 'female_age_19',
 'male_age_20',
 'female_age_20',
 'male_age_21',
 'female_age_21',
 'male_age_22',
 'female_age_22',
 'male_age_23',
 'female_age_23',
 'male_age_24',
 'female_age_24',
 'male_age_25',
 'female_age_25',
 'male_age_26',
 'female_age_26',
 'male_age_27',
 'female_age_27',
 'male_age_28',
 'female_age_28',
 'male_age_29',
 'female_age_29',


หลังจากสร้าง column รายอายุแล้ว เราต้องนำไปรวมกับ column แรกและ column ท้ายไฟล์

In [19]:
population_columns = ( 
    ["desc_ccaatt"] 
    + age_columns 
    + [ 
        "male_tdob", 
        "female_tdob", 
        "total_tdob", 
        "male_xhouse", 
        "female_xhouse", 
        "total_xhouse", 
        "male_other_nat", 
        "female_other_nat", 
        "total_other_nat", 
        "male_move", 
        "female_move", 
        "total_move", 
        "male_tot", 
        "female_tot", 
        "total_tot" 
    ] 
)

population_columns

['desc_ccaatt',
 'male_age_0',
 'female_age_0',
 'male_age_1',
 'female_age_1',
 'male_age_2',
 'female_age_2',
 'male_age_3',
 'female_age_3',
 'male_age_4',
 'female_age_4',
 'male_age_5',
 'female_age_5',
 'male_age_6',
 'female_age_6',
 'male_age_7',
 'female_age_7',
 'male_age_8',
 'female_age_8',
 'male_age_9',
 'female_age_9',
 'male_age_10',
 'female_age_10',
 'male_age_11',
 'female_age_11',
 'male_age_12',
 'female_age_12',
 'male_age_13',
 'female_age_13',
 'male_age_14',
 'female_age_14',
 'male_age_15',
 'female_age_15',
 'male_age_16',
 'female_age_16',
 'male_age_17',
 'female_age_17',
 'male_age_18',
 'female_age_18',
 'male_age_19',
 'female_age_19',
 'male_age_20',
 'female_age_20',
 'male_age_21',
 'female_age_21',
 'male_age_22',
 'female_age_22',
 'male_age_23',
 'female_age_23',
 'male_age_24',
 'female_age_24',
 'male_age_25',
 'female_age_25',
 'male_age_26',
 'female_age_26',
 'male_age_27',
 'female_age_27',
 'male_age_28',
 'female_age_28',
 'male_age_29',
 '

ตอนนี้เราได้ list ของชื่อ column ทั้งหมดแล้ว 

จำนวน column ควรประกอบด้วย 
- 1 column แรก คือ `desc_ccaatt` 
- 204 column รายอายุ เพราะมี 102 อายุ และแต่ละอายุมีชาย/หญิง 
- 15 column ท้ายไฟล์

รวมทั้งหมด 220 column

In [20]:
len(population_columns)

220

## 18. อ่านข้อมูลประชากรจาก URL พร้อมกำหนดชื่อ column 
เมื่อมีรายชื่อ column แล้ว เราสามารถอ่านข้อมูลจาก URL โดยใช้ parameter ต่อไปนี้ 
- `sep="|"` เพราะไฟล์ใช้ `|` เป็นตัวคั่น 
- `header=None` เพราะไฟล์ไม่มี header 
- `names=population_columns` เพื่อกำหนดชื่อ column เอง
- `index_col=False` เพื่อบอก `pandas` ว่าไม่ต้องนำคอลัมน์ใดในไฟล์ csv ไปใช้เป็น index ของ DataFrame
โดยเฉพาะเมื่อเรากำหนดชื่อคอลัมน์เองด้วย `names`

In [27]:
population_df = pd.read_csv( 
    text_file_url, 
    sep="|", 
    header=None,
    names=population_columns
)

population_df

,desc_ccaatt,male_age_0,female_age_0,male_age_1,female_age_1,male_age_2,female_age_2,male_age_3,female_age_3,male_age_4,...,total_xhouse,male_other_nat,female_other_nat,total_other_nat,male_move,female_move,total_move,male_tot,female_tot,total_tot
กรุงเทพมหานคร,11640,10838,14323,13274,15588,14635,15814,14567,17733,16699,...,37804,28545,66349,1333,1150,2483,2526272,2896296,5422568,NaN
เขตพระนคร,52,40,55,47,73,48,70,64,78,78,...,388,256,644,1,3,4,18723,19685,38408,NaN
แขวงพระบรมมหาราชวัง,4,2,7,2,3,4,2,5,4,3,...,17,16,33,1,0,1,1684,1104,2788,NaN
แขวงวังบูรพาภิรมย์,10,9,12,10,20,8,14,15,26,22,...,110,90,200,0,0,0,4172,4066,8238,NaN
แขวงวัดราชบพิธ,5,6,4,2,5,3,1,3,1,6,...,24,23,47,0,0,0,1218,1509,2727,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
เขตบางบอน,190,190,250,209,312,296,264,286,330,322,...,415,409,824,4,3,7,46570,51590,98160,NaN
แขวงบางบอนเหนือ,48,43,59,54,71,74,68,86,74,79,...,98,97,195,1,0,1,10654,11744,22398,NaN
แขวงบางบอนใต้,49,44,56,44,84,64,56,50,82,87,...,54,48,102,2,2,4,11062,12263,23325,NaN
แขวงคลองบางพราน,45,52,84,58,92,95,82,84,110,94,...,125,136,261,0,0,0,14303,16037,30340,NaN


จากผลลัพธ์เราจะสังเกตุเห็นว่าชื่อคอลัมน์ของข้อมูลยังคงดูผิดปกติ

เนื่องจากไฟล์นี้ไม่มีการกำหนดคอลัมน์สำหรับ `index`

`pandas` จึงเข้าใจว่าคอลัมน์แรกของไฟล์เป็น `index`

เราต้องกำหนด

```python 
index_col=False
```

เพื่อบอก `pandas` ไม่ต้องนำคอลัมน์ใดในไฟล์ csv ไปใช้เป็น `index`

In [29]:
population_df = pd.read_csv( 
    text_file_url, 
    sep="|", 
    header=None,
    index_col=False,
    names=population_columns
)

population_df

,desc_ccaatt,male_age_0,female_age_0,male_age_1,female_age_1,male_age_2,female_age_2,male_age_3,female_age_3,male_age_4,...,total_xhouse,male_other_nat,female_other_nat,total_other_nat,male_move,female_move,total_move,male_tot,female_tot,total_tot
0,กรุงเทพมหานคร,11640,10838,14323,13274,15588,14635,15814,14567,17733,...,73564,37804,28545,66349,1333,1150,2483,2526272,2896296,5422568
1,เขตพระนคร,52,40,55,47,73,48,70,64,78,...,436,388,256,644,1,3,4,18723,19685,38408
2,แขวงพระบรมมหาราชวัง,4,2,7,2,3,4,2,5,4,...,75,17,16,33,1,0,1,1684,1104,2788
3,แขวงวังบูรพาภิรมย์,10,9,12,10,20,8,14,15,26,...,62,110,90,200,0,0,0,4172,4066,8238
4,แขวงวัดราชบพิธ,5,6,4,2,5,3,1,3,1,...,6,24,23,47,0,0,0,1218,1509,2727
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
226,เขตบางบอน,190,190,250,209,312,296,264,286,330,...,1677,415,409,824,4,3,7,46570,51590,98160
227,แขวงบางบอนเหนือ,48,43,59,54,71,74,68,86,74,...,0,98,97,195,1,0,1,10654,11744,22398
228,แขวงบางบอนใต้,49,44,56,44,84,64,56,50,82,...,1677,54,48,102,2,2,4,11062,12263,23325
229,แขวงคลองบางพราน,45,52,84,58,92,95,82,84,110,...,0,125,136,261,0,0,0,14303,16037,30340


# แบบฝึกหัดท้ายบท
ให้ผู้เรียนทำแบบฝึกหัดต่อไปนี้ โดยอ่านโจทย์สถานการณ์ แล้วเขียน code เพื่อนำเข้าข้อมูลให้เหมาะสมกับงาน

## แบบฝึกหัดที่ 1: รับไฟล์ CSV จากระบบงานสังคมและเตรียม DataFrame ตั้งต้น

คุณได้รับไฟล์ logbook จากระบบงานสังคมชื่อ

```python
"msdhs_ops_mso_logbook.csv"
```

ไฟล์นี้เป็น CSV ภาษาไทย เจ้าหน้าที่แจ้งว่าเมื่อเปิดด้วยบางโปรแกรมอาจมีปัญหาภาษาไทยเพี้ยน

ให้เขียน code เพื่อนำเข้าไฟล์นี้เป็น DataFrame ชื่อ mso_df

ข้อกำหนด:

1. ใช้ pd.read_csv()
2. กำหนด encoding ให้เหมาะกับไฟล์ CSV ภาษาไทย
3. เก็บชื่อไฟล์ไว้ในตัวแปร csv_file ก่อนอ่านไฟล์
4. แสดงผล DataFrame ที่อ่านได้

In [ ]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 2: ทดลองอ่านไฟล์ขนาดใหญ่ก่อนนำเข้าจริง 
ก่อนอ่านไฟล์ logbook ทั้งหมด หัวหน้าทีมต้องการให้คุณทดลองอ่านข้อมูลเพียงบางส่วนก่อน 

เพื่อดูว่าไฟล์เปิดได้ถูกต้องหรือไม่ และภาษาไทยแสดงผลได้หรือไม่ 

ให้เขียน code อ่านไฟล์ `msdhs_ops_mso_logbook.csv` เพียง 10 แถวแรก 

ข้อกำหนด: 
1. ใช้ตัวแปร `csv_file` จากข้อก่อนหน้า 
2. ใช้ `nrows` เพื่อจำกัดจำนวนแถว 
3. เก็บผลลัพธ์ไว้ในตัวแปร `mso_preview` 
4. แสดงผล `mso_preview`

In [ ]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 3: อ่านเฉพาะข้อมูลที่เกี่ยวข้องกับการจัดทำรายงานพื้นที่ 
ทีมวิเคราะห์ต้องการจัดทำรายงานเบื้องต้นเกี่ยวกับกลุ่มเป้าหมายในแต่ละจังหวัด 

จึงไม่ต้องการอ่านทุก column จากไฟล์ logbook ให้เลือกอ่านเฉพาะ column z

ต่อไปนี้จากไฟล์ `msdhs_ops_mso_logbook.csv` 
- `รหัสครัวเรือน` 
- `วันที่สร้างครัวเรือน` 
- `จังหวัด` 
- `อายุ` 
- `เพศ` 
- `ระดับการศึกษา` 
- `อาชีพหลัก` 
- `ประเภทกลุ่มเป้าหมาย` 
- `หน่วยงานของ cm` 

ข้อกำหนด: 
1. สร้าง list ชื่อ `mso_report_columns` เพื่อเก็บรายชื่อ column ข้างต้น 
2. ใช้ `usecols` ในการอ่านเฉพาะ column เหล่านี้ 
3. เก็บผลลัพธ์ไว้ในตัวแปร `mso_report_df` 
4. แสดงผล `mso_report_df`

In [ ]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 4: ป้องกันปัญหารหัสถูกอ่านเป็นตัวเลข 
หลังจากอ่านไฟล์ logbook ทีมพบว่า column ที่เป็นรหัสไม่ควรถูกอ่านเป็นตัวเลข 

เพราะไม่ได้ใช้สำหรับการคำนวณ และอาจมีเลขศูนย์นำหน้าหรือรูปแบบเฉพาะของระบบ 

ให้อ่านไฟล์ `msdhs_ops_mso_logbook.csv` โดยเลือกเฉพาะ column ต่อไปนี้ 
- `รหัสครัวเรือน` 
- `รหัสประจำบ้าน` 
- `รหัส cm` 
- `จังหวัด` 
- `อายุ` 
- `ประเภทกลุ่มเป้าหมาย` 

ข้อกำหนด: 
1. กำหนดให้ `รหัสครัวเรือน`, `รหัสประจำบ้าน` และ `รหัส cm` ถูกอ่านเป็น string 
2. ใช้ `usecols` และ `dtype` ร่วมกัน 
3. เก็บผลลัพธ์ไว้ในตัวแปร `mso_codes_df` 
4. แสดงผล `mso_codes_df`

In [ ]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 5: เตรียมข้อมูลวันที่สำหรับการวิเคราะห์ 
ทีมต้องการใช้ข้อมูลวันที่เพื่อวิเคราะห์ต่อ เช่น ตรวจช่วงเวลาการสร้างครัวเรือนและวันที่แก้ไขข้อมูลล่าสุด 

ให้อ่านไฟล์ `msdhs_ops_mso_logbook.csv` โดยกำหนดให้ column วันที่ต่อไปนี้ถูกอ่านเป็นวันที่ 
- `วันที่แก้ไขข้อมูลล่าสุด` 
- `วันที่สร้างครัวเรือน` 

ข้อกำหนด: 
1. ใช้ `parse_dates` 
2. กำหนด encoding ให้เหมาะสม 
3. เก็บผลลัพธ์ไว้ในตัวแปร `mso_dates_df` 
4. แสดงผล `mso_dates_df`

In [ ]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 6: รับไฟล์ Excel หลาย sheet จากหน่วยงานเกษตร

คุณได้รับไฟล์ Excel ชื่อ

```python
"moac_opsmoac_fragile_farmer.xlsx"
```

ไฟล์นี้มีข้อมูลหลาย sheet จากหลายหน่วยงาน ทีมต้องการทราบก่อนว่าไฟล์นี้มี sheet อะไรบ้าง

ให้เขียน code เพื่อตรวจรายชื่อ sheet ทั้งหมดในไฟล์

ข้อกำหนด:
1. เก็บชื่อไฟล์ไว้ในตัวแปร excel_file
2. ใช้ `pd.ExcelFile()`
3. เก็บรายชื่อ sheet ไว้ในตัวแปร sheet_names
4. แสดงผล sheet_names

In [ ]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 7: อ่านข้อมูลเกษตรกรจาก sheet ที่ต้องการ

หลังจากดูรายชื่อ sheet แล้ว ทีมตัดสินใจเริ่มจาก sheet ของกรมส่งเสริมการเกษตร คือ

```python
"v_doae_fragile"
```

ให้เขียน code อ่าน sheet นี้จากไฟล์ `moac_opsmoac_fragile_farmer.xlsx`

ข้อกำหนด:
1. ใช้ `pd.read_excel()`
2. กำหนด `sheet_name="v_doae_fragile"`
3. เก็บผลลัพธ์ไว้ในตัวแปร `doae_df`
4. แสดงผล `doae_df`

In [30]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 8: อ่านเฉพาะ column สำคัญจากข้อมูลเกษตรกร 
ทีมต้องการเตรียมข้อมูลจาก sheet `v_doae_fragile` สำหรับวิเคราะห์รายได้ หนี้สิน และพื้นที่ของเกษตรกร 

โดยยังไม่ต้องอ่านทุก column ให้เลือกอ่านเฉพาะ column ต่อไปนี้ 
- `department_code` 
- `department` 
- `pid` 
- `province_code` 
- `province` 
- `amphur` 
- `tambon` 
- `is_farmer` 
- `farmer_type` 
- `income_in` 
- `income_out` 
- `debts_in` 
- `debts_out` 
- `updated_at` 

ข้อกำหนด: 
1. สร้าง list ชื่อ `farmer_columns` 
2. ใช้ `usecols=farmer_columns` 
3. กำหนดให้ `pid` และ `province_code` ถูกอ่านเป็น string 
4. เก็บผลลัพธ์ไว้ในตัวแปร `doae_selected_df` 
5. แสดงผล `doae_selected_df`

In [31]:
# เขียนคำตอบของคุณใน Cell นี้

## แบบฝึกหัดที่ 9: อ่านข้อมูลประชากรจาก URL ที่ไม่มี header

ทีมต้องการนำเข้าข้อมูลจำนวนประชากรแยกรายอายุของจังหวัดกรุงเทพมหานคร ปี 2568 จาก URL ต่อไปนี้

```python
"https://stat.bora.dopa.go.th/new_stat/file/6812/6812cc10.txt"
```

ไฟล์นี้เป็นไฟล์ text ที่
- อยู่บน URL
- ใช้ `|` เป็นตัวคั่นข้อมูล
- ไม่มี `header`
- มีจำนวน column มาก

ให้ทดลองอ่านข้อมูลเพียง 5 แถวแรกก่อน เพื่อดูว่าอ่านไฟล์ได้หรือไม่

ข้อกำหนด:
1. เก็บ URL ไว้ในตัวแปร population_url
2. ใช้ pd.read_csv()
3. กำหนด sep="|"
4. กำหนด header=None
5. ใช้ nrows=5
6. เก็บผลลัพธ์ไว้ในตัวแปร population_preview
7. แสดงผล population_preview

## แบบฝึกหัดที่ 10: สร้าง workflow นำเข้าข้อมูลประชากรจาก URL ให้พร้อมใช้ต่อ

ทีมต้องการนำเข้าข้อมูลประชากรจาก URL ให้มีชื่อ column ที่อ่านเข้าใจได้ เพื่อเตรียมไว้ใช้ตรวจสอบและวิเคราะห์ต่อในบทที่ 9

จากโครงสร้างข้อมูล ไฟล์นี้มีลำดับ column ดังนี้

1. column แรกคือรายละเอียดพื้นที่ ให้ใช้ชื่อ `desc_ccaatt`
2. ข้อมูลประชากรรายอายุ ตั้งแต่อายุ 0 ถึง 101 โดยแต่ละอายุมี 2 column คือชายและหญิง
3. column ท้ายไฟล์เป็นข้อมูลกลุ่มพิเศษและยอดรวม

ให้เขียน workflow สำหรับนำเข้าข้อมูลนี้เป็น DataFrame ชื่อ `population_df`

ข้อกำหนด:

1. สร้าง list ชื่อ `age_columns` ด้วย loop โดยให้มีชื่อ column ดังนี้  
   - `male_age_0`, `female_age_0`
   - ...
   - `male_age_101`, `female_age_101`

2. สร้าง list ชื่อ `ending_columns` สำหรับ column ท้ายไฟล์ดังนี้

```python
[
    "male_tdob",
    "female_tdob",
    "total_tdob",
    "male_xhouse",
    "female_xhouse",
    "total_xhouse",
    "male_other_nat",
    "female_other_nat",
    "total_other_nat",
    "male_move",
    "female_move",
    "total_move",
    "male_tot",
    "female_tot",
    "total_tot"
]
```

3. รวมชื่อ column ทั้งหมดไว้ใน list ชื่อ population_columns

population_columns = ["desc_ccaatt"] + age_columns + ending_columns

4. ใช้ `pd.read_csv()` อ่านข้อมูลจาก population_url โดยกำหนด
- sep=`"|"`
- header=None
- names=population_columns
- dtype={"desc_ccaatt": "string"}

5. เก็บผลลัพธ์ไว้ในตัวแปร population_df
6. จากนั้นสร้าง DataFrame อีกชุดชื่อ population_total_df โดยอ่านเฉพาะ column ต่อไปนี้
- desc_ccaatt
- male_tot
- female_tot
- total_tot

In [ ]:
# เขียนคำตอบของคุณใน Cell นี้